worker/upload.go

```diff
@@ -0,0 +1,836 @@
+		sectors     []object.Sector
+	}
+
+	average struct {
+		mu             sync.Mutex
+		avg            float64
+		pts            [1e3]float64
+		cnt            uint64
+		updateInterval time.Duration
+		lastUpdate     time.Time
+	}
+)
+
+func (u *uploader) newQueue(c api.ContractMetadata) *uploadQueue {
+	return &uploadQueue{
+		u:  u,
+		hp: u.w,
+		rl: u.w,
+
+		fcid:       c.ID,
+		hk:         c.HostKey,
+		siamuxAddr: c.SiamuxAddr,
+
+		queue:        make([]*uploadJob, 0),
+		queueChan:    make(chan struct{}),
+		queueUploads: make(map[uploadID]struct{}),
+
+		statsSpeed:        newAverage(defaultAvgUpdateInterval),
+		statsEstimateDiff: newAverage(defaultAvgUpdateInterval),
+		stopChan:          make(chan struct{}),
+	}
+}
+
+func newAverage(updateInterval time.Duration) *average {
+	return &average{
+		updateInterval: updateInterval,
+		pts:            [1e3]float64{},
+	}
+}
+
+func (w *worker) initUploader() {
+	if w.uploader != nil {
+		panic("uploader already initialized") // developer error
+	}
+
+	w.uploader = &uploader{
+		statsOverdrive: newAverage(defaultAvgUpdateInterval),
+		statsStopChan:  make(chan struct{}),
+
+		w:    w,
+		pool: make([]*uploadQueue, 0),
+	}
+}
+
+func (u *uploader) Stop() {
+	u.poolMu.Lock()
+	defer u.poolMu.Unlock()
+	for _, q := range u.pool {
+		close(q.stopChan)
+	}
+	close(u.statsStopChan)
+}
+
+func (u *uploader) blockHeight() uint64 {
+	u.poolMu.Lock()
+	defer u.poolMu.Unlock()
+	return u.bh
+}
+
+func (u *uploader) newUpload() uploadID {
+	var id uploadID
+	frand.Read(id[:])
+	return id
+}
+
+func (u *uploader) finishUpload(id uploadID) {
+	u.poolMu.Lock()
+	defer u.poolMu.Unlock()
+	for _, q := range u.pool {
+		q.finish(id)
+	}
+}
+
+func (u *uploader) upload(ctx context.Context, r io.Reader, contracts []api.ContractMetadata, rs api.RedundancySettings, bh uint64) (o object.Object, used map[types.PublicKey]types.FileContractID, err error) {
+	// add tracing
+	ctx, span := tracing.Tracer.Start(ctx, "uploader.upload")
+	defer func() {
+		if err != nil {
+			span.RecordError(err)
+		}
+		span.End()
+	}()
+
+	// sanity check redundancy
+	if rs.TotalShards > len(contracts) {
+		err := errors.New("not enough contracts to meet redundancy")
+		span.RecordError(err)
+		return object.Object{}, nil, err
+	}
+
+	// refresh contracts
+	u.updatePool(contracts, bh)
+
+	// keep track of used hosts
+	used = make(map[types.PublicKey]types.FileContractID)
+
+	// build contracts map
+	h2c := make(map[types.PublicKey]types.FileContractID)
+	for _, c := range contracts {
+		h2c[c.HostKey] = c.ID
+	}
+
+	// initialize the encryption key
+	o.Key = object.GenerateEncryptionKey()
+	cr := o.Key.Encrypt(r)
+
+	for {
+		proceed, err := func() (bool, error) {
+			// ensure the upload ctx is cancelled
+			ctx, cancel := context.WithCancel(ctx)
+			defer cancel()
+
+			// create span
+			ctx, span := tracing.Tracer.Start(ctx, "uploader.slab")
+			defer span.End()
+
+			// initialize upload
+			id := u.newUpload()
```

Comment:
Naming-wise it's slightly confusing that `newUpload` is called in every iteration. Also the fact that we need to call `newUpload` and `finishUpload`. Can't we restructure it so that we only need to call `uploadSectors` in a loop?